src/main/flight/gps_rescue.c

```diff
@@ -404,12 +403,14 @@ static void rescueAttainPosition()
     // D component
     float throttleD = ((altitudeError - previousAltitudeError) / sampleIntervalNormaliseFactor);
     previousAltitudeError = altitudeError;
-    // simple first order filter on derivative with k = 0.3
-    throttleD = 0.7 * previousThrottleD +  0.3f * throttleD;
+    throttleD = MAX(throttleD, 0.0f); // for altitude control, D is used as an early upwards push only
+    throttleD = 0.5 * previousThrottleD +  0.5f * throttleD; // simple first order filter on derivative with k = 0.5
```

Comment:
Choosing a constant k value will get you different cutoff frequencies, depending on the GPS update rate.
Better use `k = pt1FilterGain()` to get the same lowpass behavior on every GPS update rate.

Reply:
Yes, good point.  However this is primarily a 'step smoother'.  I haven't yet seen faster data so I don't know if it is more or less noisy.  I am hoping it is no more noisy.  If there is no more noise at the faster rate, the same amount of step to step smoothing should be sufficient, and would cause less delay.
This is perhaps something we can deal with when we find a GPS with a higher packet rate.

Reply:
Yes, faster rates will make it more noisy. You need to compensate for the rate with `k = pt1FilterGain()` to get a consistent noise filtration on all rates.

Reply:
Thanks for your thoughts on this.

The optimal kind of filtering depends on the nature of the noise, how tolerant we are of delay, and the harm caused by the noise.   

Our data points come in only at 5Hz.  Additional delay is a significant concern.  

The corrections we apply are typically relatively small.  

For throttle, each step results in an immediate rise to a new throttle value.  It is held for 200ms before the next change.  We need a second derivative of height. I am amazed to find that the noise problem isn't nearly as bad as I expected.  Where there are steps in PIDsum, it results in a throttle change only every 200ms.  The absolute magnitude of the throttle steps is not great.  From a motor heat / energy wastage perspective, strong smoothing isn't going to help much; conversely, a lot of delay will be a significant problem.

I've found that a significant element of D and Jerk noise in the altitude controller is of the sequential up/down stepping type.  This responds much better to a two-point moving average than to a PT1 type filter.  That's why the altitude filter is now a 2-point moving average of the sum of jerk and D - not a PT1 type filter at all.  This provides better smoothing for about the same delay, because the noise itself is of the type that responds best to a moving average.

This is a graphic showing the outcome with too much D+Jerk (causing the oscillation).  There is no meaningful problem with throttle P, and all the steps in D are sequential.  This is with the moving average on D and no smoothing at all on P.  I think this is a perfectly adequate result.

![Screen Shot 2022-05-25 at 16 00 19](https://user-images.githubusercontent.com/11737748/170190739-ec0e7623-baa5-428b-ba5b-bf286da37da7.jpg)

For velocity, the control value is itself the derivative of distance.  It isn't surprising that its derivative is noisy.  Also, each step changes requested angle, leading to a spike in the gyro and motor traces as the PID controller implement the change in angle.  These spikes are best addressed by smoothing the steps in the requested pitch angle with say a PT3 filter.  However, the changes in pitch do not result in significant motor noise, because most of the time the flight angle doesn't change much, if at all. 

Check out this image of velocity P and D.  The gyro pitch signal change is small, and the motor changes to achieve those steps are also small.  Note that noise on un-filtered P exists (sequential up-down steps) and also there is some on D (after a PT1 filter), but neither are really bad.  I then have a PT1 filter on the sum of both.  The little pulses on gyro are typically in the same direction and of similar magnitude, reflecting very little noise in the final output.  I will make some special debugs and probably will change the final PT1 to a two point moving average.  It will do better than the PT1 for sequential up/down changes, and there are a lot of them.

![Screen Shot 2022-05-25 at 15 46 44](https://user-images.githubusercontent.com/11737748/170189987-aba8d3c1-c32c-42cc-a40b-b7b1839553dc.jpg)

I think I need to do some more work on the velocity filtering, and will optimise it to achieve a good result with the least delay. That will need some specific logging and an analysis of the nature of the noise.

I am very happy with altitude/throttle smoothing using the moving average.

I understand that some GPS units report at 1hz. For them no filtering at all would likely be best.  I don't think there is anything in-between 5hz and 1hz but I have heard of some units running 10hz, though I haven't seen any, yet.  
Until I see signals at other data rates, I won't know the best way to filter that data.  When I do get a 10hz log, then we will know if we need to do something more complex.

Reply:
Thanks, that's great!
The main thing is that I don't know for sure what the noise picture is like at 1Hz, or at 10Hz, which I think are the only two alternate frequencies.
Until I see data from GPS at those frequencies, I don't know what is needed.
I expect that at 1Hz there should be no filtering at all.
At 10Hz I need to see the noise picture first.  I honestly don't know what it is like.

Reply:
> At 10Hz I need to see the noise picture first. I honestly don't know what it is like.

This is another argument to make the k value change with GPS rate because without it you will have twice as much noise at 10Hz than at 5Hz at high frequencies. With adjusted k value by `pt1FilterGain()` the signal will be as clean on both rates. On 1Hz the filtered signal is basically unfiltered because of the high cutoff in comparison to sample rate so we are good to use `pt1FilterGain()`.

Reply:
Yes that makes sense, but it assumes there will be a lot more noise in the 10Hz GPS.  They may use better algorithms and not have significantly worse noise.  But, in any case, your solution is a robust one, and easy to implement.  Thank you.